File: Part 2 - Regression/Section 4 - Simple Linear Regression/simple_linear_regression_1.r
# Simple Linear Regression

# Importing DataSet
dataset = read.csv('Salary_Data.csv')

# Splitting dataset into training set and test set
library(caTools)
set.seed(123)
split = sample.split(dataset$Salary, SplitRatio = 2/3)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# We don't need feature scaling here as simple linear regression in R takes care of it.
# Fitting simple linear regression to Training Set

# Salary ~ YearsExperience means Salary is proportional to YearsExperience
regressor = lm(formula = Salary ~ YearsExperience,
               data = training_set)
# To get some information about simple linear model, in console type summary(regressor) 
# In the summary, in coefficients, if there are no stars, it means no statistical significance. If there are 3 stars, it means that the model has high statistical significance
# P value is another indicator of statistical significance. Lower the p value, the higher the impact of independent variable on the dependent variable. Usually, if p value is lower than 5%, the independent variable is highly significant. Anything lower is less significant.

# Predicting Test Set results
y_pred = predict(regressor, newdata = test_set)

# Visualising the Training Set results
# install.packages('ggplot2')
library(ggplot2)
ggplot() +
  # plot observation points
  geom_point(aes(x = training_set$YearsExperience, y = training_set$Salary),
             colour = 'red') +
  # since we want to take predicted salaries, y cannot be of training_set
  geom_line(aes(x = training_set$YearsExperience, y = predict(regressor, newdata = training_set)),
            colour = 'blue') +
  ggtitle('Salary vs Experience (Training Set)') +
  xlab('Years of Experience') +
  ylab('Salary')

# Visualising the Test Set results
ggplot() +
  # plot observation points
  geom_point(aes(x = test_set$YearsExperience, y = test_set$Salary),
             colour = 'red') +
  # since we want to take predicted salaries, y cannot be of training_set
  geom_line(aes(x = training_set$YearsExperience, y = predict(regressor, newdata = training_set)),
            colour = 'blue') +
  ggtitle('Salary vs Experience (Test Set)') +
  xlab('Years of Experience') +
  ylab('Salary')

File: Part 2 - Regression/Section 4 - Simple Linear Regression/simple_linear_regression_1.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Oct 23 08:43:41 2017

@author: shreyas
"""
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Salary_Data.csv')
# importing independent variable 'x'
# : = all rows, :-1 = except last column, take all rows except last column
X = dataset.iloc[:, :-1].values
# importing dependent variable vector 'y'
# : = all rows, 1 = second column as index starts from 0
y = dataset.iloc[:, 1].values
# X size is (30, 1) but y is (30,). It tells that X is matrix that has 1 column and since y doesn't have anything, its a vector.
# X is the matrix of (feature) independent variable and y is a vector of dependent variable

# Splitting dataset into training set and test set
from sklearn.cross_validation import train_test_split
# Usually test set !> 0.4. test set + train set = 1. random_state is not required
# Here test set is 1/3 the size of the data and training set would be 2/3
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=1/3, random_state=0)

# Simple Linear Regression in Python takes care of feature scaling, so we don't have to worry about that.

# Fitting Simple Linear Regression to the Training set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
# fitting regressor object to training set
# fits the line and returns the slope
regressor.fit(X_train, y_train)

# Predicting the test set results
y_pred = regressor.predict(X_test)

# Visualising the training set results
# observation points (real values)
plt.scatter(X_train, y_train, color = 'red')
# regression line
plt.plot(X_train, regressor.predict(X_train), color = 'blue')
plt.title('Salary vs Experience (Training Set)')
plt.xlabel('Years of Experience')
plt.ylabel('Salary')
plt.show()

# Visualising the test set results
plt.scatter(X_test, y_test, color = 'red')
# since regressor is trained on the training set, regression line is already determined
plt.plot(X_train, regressor.predict(X_train), color = 'blue')
plt.title('Salary vs Experience (Test Set)')
plt.xlabel('Years of Experience')
plt.ylabel('Salary')
plt.show()